primitives: Update supply curve

jsdanielh opened the pull request:
Update supply curve to reflect the community voted curve which is the same used in the PoW chain (ideal curve).

This closes #2380.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
primitives: Update supply curve

Update supply curve to reflect the community voted curve which is
the same used in the PoW chain.

This closes #2380.

sisou commented on primitives/src/policy.rs:
```diff
@@ -121,13 +121,13 @@ impl Policy {
-    pub const INITIAL_SUPPLY_VELOCITY: f64 = 875.0;
+    /// Supply_velocity (t) = Initial_supply_velocity * 2^(- Supply_decay * t)
+    /// Where t is the time in milliseconds since the genesis block.
+    pub const INITIAL_SUPPLY_VELOCITY: f64 = 0.10594132556065439;
```
Is this a number that is dependent on the PoS genesis block timestamp? Or can we fix it to a stable timestamp, e.g. PoW genesis timestamp?

jsdanielh replied:
Yes, if we want to 'continue' the curve, we would need to use the PoW timestamp. Will make a change to use it.

jsdanielh replied:
Done

hrxi commented on primitives/src/policy.rs:
```diff
@@ -502,7 +502,8 @@ impl Policy {
         let supply = genesis_supply
-            + (Self::INITIAL_SUPPLY_VELOCITY / Self::SUPPLY_DECAY * (1.0 - exp(exponent))) as u64;
+            + (Self::INITIAL_SUPPLY_VELOCITY / Self::SUPPLY_DECAY
+                * (1.0 - f64::powf(2.0, exponent))) as u64;
```
We can't use the `f64::powf` function here, since this is consensus-relevant and https://doc.rust-lang.org/std/primitive.f64.html#method.powf says:
> The precision of this function varies by platform and Rust version.

jsdanielh replied:
Ok, I added `pow` to our math utils and changed the function to use it.

jsdanielh commented:
Numbers for the regressed curved were calculated using: https://github.com/jsdanielh/supply-curve

The result is the following:
![image](https://github.com/nimiq/core-rs-albatross/assets/2374448/a48f6e45-9057-458e-85a9-3260fc1d119d)

jsdanielh committed:
primitives: Change the `supply_at` function to not require genesis TS

Change the `supply_at` function in the policy module within the
primitives subcrate to depend only on the PoW genesis timestamp
defined as a constant such that the supply _follows_ the PoW supply
curve as a continuation. This also changes the signature of this
function since the PoS genesis timestamp is no longer required.

sisou commented:
@jsdanielh but careful now with the mix of parameters. You are using the PoW timestamp, but the PoS supply. Should it use the PoW supply, to start the curve at the right point?

Also, the PoW supply was based on _block number_. The block numbers have diverged slightly from the ideal timeline, I don't remember if we are _early_ or _late_. So using the "idealized" curve from the PoS genesis will either skip a few block rewards, or pay them out double. Just something to think about.

hrxi commented on primitives/src/policy.rs:
```diff
@@ -136,6 +136,9 @@ impl Policy {
     pub const HISTORY_CHUNKS_MAX_SIZE: u64 = 25 * 1024 * 1024;
 
+    /// The PoW genesis timestamp.
+    pub const POW_GENESIS_TIMESTAMP: u64 = 1523727000;
```
What's the unit here? It looks like a timestamp using seconds as resolution, but I think we need milliseconds.

jsdanielh replied:
Yeah, you're right

jsdanielh commented:
> @jsdanielh but careful now with the mix of parameters. You are using the PoW timestamp, but the PoS supply. Should it use the PoW supply, to start the curve at the right point?

I don't think so. The genesis supply is the initial factor in the supply curve and just sets an offset on which supply we start with. The rate of decay in the other hand needs to be preserved since it is exponentially decreasing in time and needs to consider that the supply rate already decayed (6+ years of PoW life).

jsdanielh committed:
primitives: Use a power function with the same precision

Use a power function with a the same precision in any platform when
calculating the supply in the policy module.

sisou commented:
You are right that the genesis supply has no influence on the batch rewards (as for that, the increase of `supply_at` between two macro block timestamps is compared, so the initial supply added to both values cancels out).

But if you wanted to compute the actual supply at a certain point in time, using the PoS genesis supply generates a way-too-high number, as the graph is moved up on the Y-axis by the supply difference between PoW genesis and PoS genesis.

jsdanielh committed:
blockchain: Update test for new supply curve

Update `it_can_create_batch_finalization_inherents` test for new
supply curve.

sisou commented on blockchain/tests/inherents.rs:
```diff
@@ -47,7 +47,7 @@ fn it_can_create_batch_finalization_inherents() {
         block_number: Policy::macro_block_of(2).unwrap(),
         round: 0,
-        timestamp: blockchain.state.election_head.header.timestamp + 1,
+        timestamp: blockchain.state.election_head.header.timestamp + 20000,
```
Any reason for this 20s?

jsdanielh replied:
The test has assumptions in terms of the number of reward being paid which with the new supply curve don't hold for such a short timestamp difference.

jsdanielh committed:
blockchain: Fix panic when the reward is less than the number of slots

Fix panic when the reward is less than the number of slots. In this
case the reward is zero but the remainder should still be randomly
distributed among the elegible slots. For this, the filtering of
zero value reward transactions is done at the end.

hrxi commented on utils/src/math/mod.rs:
```diff
@@ -1,3 +1,5 @@
+mod pow;
 
 pub use exp::exp;
+pub use pow::pow;
```
You should be able to remove the `exp` implementation.

jsdanielh replied:
Unfortunately it is still used to calculate batch delay penalties

hrxi reviewed (commented):
I think there will be a discontinuity if we simply base it on the time since the PoW genesis block as the PoW rewards are not based on time AFAIK. Is that a problem?

jsdanielh commented:
> I think there will be a discontinuity if we simply base it on the time since the PoW genesis block as the PoW rewards are not based on time AFAIK. Is that a problem?

We're basing it in the _ideal_ PoW emission curve based on the ideal expected number of blocks per year. Having said that, indeed this is a subject of discussion.

hrxi commented on primitives/src/policy.rs:
```diff
@@ -490,19 +493,20 @@ impl Policy {
         let supply = genesis_supply
-            + (Self::INITIAL_SUPPLY_VELOCITY / Self::SUPPLY_DECAY * (1.0 - exp(exponent))) as u64;
+            + (Self::INITIAL_SUPPLY_VELOCITY / Self::SUPPLY_DECAY * (1.0 - pow(2.0, exponent)))
+                as u64;
```
Looking at this again, this looks wrong.

This function returns more than the PoS `genesis_supply` at the `current_time` of the PoS genesis block.

hrxi replied:
Since I'm looking at simplifying the formula anyway, I'll probably also fix that part.

sisou replied:
That is what [I commented above](https://github.com/nimiq/core-rs-albatross/pull/2682#issuecomment-2188872842):
> But if you wanted to compute the actual supply at a certain point in time, using the PoS genesis supply generates a way-too-high number, as the graph is moved up on the Y-axis by the supply difference between PoW genesis and PoS genesis.

sisou replied:
If you move the start of the curve to the PoW genesis, you also need to move the genesis supply there. 

_Or_, if you want to use the PoS genesis supply, you need to start the curve with the initial velocity it will have at the PoS genesis time, which is not yet fixed, so the velocity cannot be fixed yet.
